File: backend/cart-service/src/main/java/com/buy02/cart/service/CartService.java
package com.buy02.cart.service;

import com.buy02.cart.DTOs.AddToCartRequest;
import com.buy02.cart.DTOs.CartResponse;
import com.buy02.cart.DTOs.UpdateCartItemRequest;

public interface CartService {

    CartResponse getCart(String userId);

    CartResponse addToCart(AddToCartRequest request, String userId);

    CartResponse updateItemQuantity(String productId, UpdateCartItemRequest request, String userId);

    CartResponse removeCartItem(String productId, String userId);

    void clearCart(String userId);
}

File: frontend/src/app/core/services/cart.service.ts
import { Injectable, inject } from "@angular/core";
import { HttpClient } from "@angular/common/http";
import { Observable } from "rxjs";

import { API } from "@core/config/api.config";
import {
  AddToCartItem,
  Cart,
  UpdateCartItem,
} from "@core/models/cart.model";

@Injectable({ providedIn: "root" })
export class CartService {
  private readonly http = inject(HttpClient);

  getCart(): Observable<Cart> {
    return this.http.get<Cart>(API.base + API.carts.root);
  }

  addToCart(body: AddToCartItem): Observable<Cart> {
    return this.http.post<Cart>(API.base + API.carts.root, body);
  }

  updateItemQuantity(
    productId: string,
    quantity: number,
  ): Observable<Cart> {
    return this.http.put<Cart>(API.base + API.carts.item(productId), {
      quantity,
    } satisfies UpdateCartItem);
  }

  removeCartItem(productId: string): Observable<Cart> {
    return this.http.delete<Cart>(API.base + API.carts.item(productId));
  }

  clearCart(): Observable<void> {
    return this.http.delete<void>(API.base + API.carts.root);
  }
}
